src/controller/editingcontroller.js

```diff
@@ -83,36 +83,62 @@ export default class EditingController {
 			viewSelection: this.view.selection
 		} );
 
-		// Convert changes in model to view.
-		this.listenTo( this.model, 'change', ( evt, type, changes ) => {
-			this.modelToView.convertChange( type, changes );
-		}, { priority: 'low' } );
-
-		// Convert model selection to view.
-		this.listenTo( this.model, 'changesDone', () => {
-			const selection = this.model.selection;
+		// Model differ object. It's role is to buffer changes done on model and then calculates a diff of those changes.
+		// The diff is then passed to model conversion dispatcher which generates proper events and kicks-off conversion.
+		const modelDiffer = new ModelDiffer();
 
-			this.modelToView.convertSelection( selection );
-			this.view.render();
-		}, { priority: 'low' } );
+		// Before an operation is applied on model, buffer the change in differ.
+		this.listenTo( this.model, 'operation', ( evt, operation ) => {
+			modelDiffer.bufferOperation( operation );
+		} );
 
-		// Convert model markers changes.
+		// Buffer marker changes.
+		// This is not covered in buffering operations because markers may change outside of them (when they
+		// are modified using `document.markers` collection, not through `MarkerOperation`).
 		this.listenTo( this.model.markers, 'add', ( evt, marker ) => {
-			this.modelToView.convertMarker( 'addMarker', marker.name, marker.getRange() );
+			// Whenever a new marker is added, buffer that change.
+			modelDiffer.bufferMarkerChange( marker.name, null, marker.getRange() );
+
+			// Whenever marker changes, buffer that.
+			marker.on( 'change', ( evt, oldRange ) => {
+				modelDiffer.bufferMarkerChange( marker.name, oldRange, marker.getRange() );
+			} );
 		} );
 
 		this.listenTo( this.model.markers, 'remove', ( evt, marker ) => {
-			this.modelToView.convertMarker( 'removeMarker', marker.name, marker.getRange() );
+			// Whenever marker is removed, buffer that change.
+			modelDiffer.bufferMarkerChange( marker.name, marker.getRange(), null );
 		} );
 
-		// Convert view selection to model.
+		// When all changes are done, get the model diff containing all the changes and convert them to view and then render to DOM.
+		this.listenTo( this.model, 'changesDone', () => {
+			// First, before the view is changed, remove from the view markers which has changed.
+			this.modelToView.removeMarkers( modelDiffer.getMarkersToRemove() );
+
+			// Calculate a diff between old model and new model. Then convert those changes to the view.
+			this.modelToView.convertBufferedChanges( modelDiffer.getChanges() );
+
+			// After the view is updated, convert markers which has changed.
+			this.modelToView.addMarkers( modelDiffer.getMarkersToAdd() );
```

Comment:
I like how this code looks like, but we could consider calling `removeMarkers` and `addMarkers ` inside `convertBufferedChanges` since you should not call `convertBufferedChanges` whitout them.

Reply:
However I'm not sure. This code tell very well what is going on.

Reply:
I am for moving `removeMarkers` and `addMarkers` to `ModelConversionDispatcher`. You are right that there is only one way to process this conversion, and you never want to fire one of those function without firing other ones, and exactly in this order.